interfaces/builtin: add network-status interface

xavi-garcia-mena opened the pull request:
This branch adds an interface to allow using the connectivity-api dbus interface.
The goal is to have the interface ready for those applications serving and consuming network status from a easy to use interface.
The connectivity-api is now included as part of the indicator-network. 
The final goal of this interface is to have the service of the connectivity-api and clients running in IoT devices running Ubuntu Core, not within the full unity8 stack.

mvo5 commented:
A bit of a meta-question - if its just about observing the network connectivity status, shouldn't we just fold this into the default network interface? @jdstrand, what do you think?

xavi-garcia-mena commented:
I was thinking exactly the same. 
It is probably a good idea to add the rules in this interface to the already existing network interface.

jdstrand commented:
@mvo5 - we used to do that on Touch. The problem with that on snappy is that the connectivity-api is not part of the core snap, so making it part of the implicit core network interface feels wrong. Put another way, snaps wanting to use connectivity-api will need to 'plugs: [connectivity-api]' to have the slot implementation installed anyway (when we start installing/prompting for missing slots anyway), so the act of doing that will give the snap these rules.

jdstrand commented on interfaces/builtin/connectivity_observe.go:
```diff
@@ -0,0 +1,156 @@
+	"github.com/snapcore/snapd/interfaces"
+)
+
+var connectivityObservePermanentSlotAppArmor = []byte(`
```
Please use 'const' instead of 'var' here and elsewhere for security policy. See https://github.com/snapcore/snapd/pull/2946 for the types of changes you need.

xavi-garcia-mena replied:
Done

jdstrand commented on interfaces/builtin/basedeclaration.go:
```diff
@@ -209,6 +209,11 @@ slots:
+    allow-installation:
+      slot-snap-type:
+        - app
+        - core
```
This should also have:
```
  deny-connection: true
```

so that a snap declaration is required for the slot implementation to have plugging snaps connect to it. We want to allow auto-connection once the snap declaration is granted, so leaving that out is fine.

xavi-garcia-mena replied:
Done

jdstrand commented on interfaces/builtin/connectivity_observe.go:
```diff
@@ -0,0 +1,156 @@
+   path=/org/freedesktop/DBus
+   interface=org.freedesktop.DBus
+   member={Request,Release}Name
+   peer=(name=org.freedesktop.DBus),
```
Please adjust to be:
```
peer=(name=org.freedesktop.DBus, label=unconfined),
```

xavi-garcia-mena replied:
Done

jdstrand commented on interfaces/builtin/connectivity_observe.go:
```diff
@@ -0,0 +1,156 @@
+		snippet := []byte(connectivityObserveConnectedSlotAppArmor)
+		old := []byte("###SLOT_SECURITY_TAGS###")
+		var new []byte
+		new = slotAppLabelExpr(slot)
```
This should be:
```
new = plugAppLabelExpr(plug)
```

xavi-garcia-mena replied:
Done

jdstrand commented on interfaces/builtin/connectivity_observe.go:
```diff
@@ -0,0 +1,156 @@
+		var new []byte
+		new = slotAppLabelExpr(slot)
+		snippet = bytes.Replace(snippet, old, new, -1)
+		return snippet, nil
```
No substitutions here. Please use simply:
```
case interfaces.SecurityAppArmor:
        return connectivityObservePermanentSlotAppArmor, nil
```

How this works is that Permanent policy is what is granted to allow the thing to run at all, and Connected policy is what is granted to allow snaps to communicate. As such, connectivityObservePermanentSlotAppArmor has no substitution rules. connectivityObserveConnectedSlotAppArmor has substitution rules for the peer plug's labels and connectivityObserveConnectedPlugAppArmor has substitution rules for the peer slot's labels.

xavi-garcia-mena replied:
Done

jdstrand commented on interfaces/builtin/connectivity_observe.go:
```diff
@@ -0,0 +1,156 @@
+		snippet := []byte(connectivityObserveConnectedPlugAppArmor)
+		old := []byte("###PLUG_SECURITY_TAGS###")
+		var new []byte
+		new = slotAppLabelExpr(slot)
```
This should use "###SLOT_SECURITY_TAGS###".

xavi-garcia-mena replied:
Done

jdstrand commented on interfaces/builtin/connectivity_observe.go:
```diff
@@ -0,0 +1,156 @@
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		snippet := []byte(connectivityObserveConnectedSlotAppArmor)
+		old := []byte("###SLOT_SECURITY_TAGS###")
```
This should be '###PLUG_SECURITY_TAGS###' since you want to use this in connectivityObserveConnectedSlotAppArmor:

xavi-garcia-mena replied:
Done

jdstrand commented on interfaces/builtin/connectivity_observe.go:
```diff
@@ -0,0 +1,156 @@
+dbus (receive)
+    bus=system
+    path=/com/ubuntu/connectivity1/NetworkingStatus{,/**}
+    interface=org.freedesktop.DBus.*,
```
Please add:
```
peer=(label=###PLUG_SECURITY_TAGS###),
```

Then adjust

xavi-garcia-mena replied:
Done

jdstrand commented on interfaces/builtin/connectivity_observe.go:
```diff
@@ -0,0 +1,156 @@
+dbus (receive, send)
+    bus=system
+    interface=com.ubuntu.connectivity1.NetworkingStatus{,/**}
+    path=/com/ubuntu/connectivity1/NetworkingStatus,
```
Please add:
```
peer=(label=###SLOT_SECURITY_TAGS###),
```

xavi-garcia-mena replied:
Done

jdstrand commented on interfaces/builtin/connectivity_observe.go:
```diff
@@ -0,0 +1,156 @@
+dbus (send)
+    bus=system
+    path=/com/ubuntu/connectivity1/NetworkingStatus{,/**}
+    interface=org.freedesktop.DBus.*,
```
Please add:
```
peer=(label=###SLOT_SECURITY_TAGS###),
```

xavi-garcia-mena replied:
Done

jdstrand commented on interfaces/builtin/connectivity_observe.go:
```diff
@@ -0,0 +1,156 @@
+<policy context="default">
+    <deny own="com.ubuntu.connectivity1.NetworkingStatus"/>
+
+    <deny send_destination="com.ubuntu.connectivity1.NetworkingStatus"/>
```
Shouldn't this also be:
```
<allow send_destination="com.ubuntu.connectivity1.NetworkingStatus"/>
```

I thought the point of the connectivity-api is that it is safe from anyone to use? What you have means only root can use it.

xavi-garcia-mena replied:
You are absolutely right.
Done.

jdstrand commented on interfaces/builtin/connectivity_observe_test.go:
```diff
@@ -0,0 +1,83 @@
+	// connected slots have a non-nil security snippet for apparmor
+	snippet, err := s.iface.ConnectedSlotSnippet(s.plug, s.slot, interfaces.SecurityAppArmor)
+	c.Assert(err, IsNil)
+	c.Assert(snippet, Not(IsNil))
```
Can you verify that something from your ConnectedSlot policy was added? Eg:
```
c.Check(string(snippet), testutil.Contains, "peer=(label=\"snap.connectivity-observe.*\"")
```

xavi-garcia-mena replied:
I've added the following:
	c.Check(string(snippet), testutil.Contains, "interface=org.freedesktop.DBus.*")
	c.Check(string(snippet), testutil.Contains, "peer=(label=\"snap.client.*\"")

I'm not sure about this as jdstrand was suggesting: label=\"snap.connectivity-observe.*, but then running the tests I saw the label was different. 
Could this be an error from my side?

jdstrand replied:
No, your tests look great now. Since you are mocking "client" and "server" you are using the right labels in your tests.

jdstrand commented on interfaces/builtin/connectivity_observe_test.go:
```diff
@@ -0,0 +1,83 @@
+	// slots have a permanent non-nil security snippet for apparmor
+	snippet, err = s.iface.PermanentSlotSnippet(s.slot, interfaces.SecurityAppArmor)
+	c.Assert(err, IsNil)
+	c.Assert(snippet, Not(IsNil))
```
Can you verify that something from your PermanentSlot policy was added? Eg:
```
c.Check(string(snippet), testutil.Contains, "dbus (bind)")
```

xavi-garcia-mena replied:
I've added:
	c.Check(string(snippet), testutil.Contains, "dbus (bind)")
	c.Check(string(snippet), testutil.Contains, "name=\"com.ubuntu.connectivity1.NetworkingStatus\"")

jdstrand commented on interfaces/builtin/connectivity_observe_test.go:
```diff
@@ -0,0 +1,83 @@
+	// slots have a permanent non-nil security snippet for dbus
+	snippet, err = s.iface.PermanentSlotSnippet(s.slot, interfaces.SecurityDBus)
+	c.Assert(err, IsNil)
+	c.Assert(snippet, Not(IsNil))
```
Can you verify that something from your PermanentSlot policy was added? Eg:
```
c.Check(string(snippet), testutil.Contains, "<policy user=\"root\">")
```

xavi-garcia-mena replied:
I've added:
	c.Check(string(snippet), testutil.Contains, "<policy user=\"root\">")
	c.Check(string(snippet), testutil.Contains, "<allow send_destination=\"com.ubuntu.connectivity1.NetworkingStatus\"/>")

jdstrand commented on interfaces/builtin/connectivity_observe.go:
```diff
@@ -0,0 +1,156 @@
+#include <abstractions/dbus-strict>
+
+# Allow all access to NetworkingStatus service
+dbus (receive, send)
```
Should this be only 'send'? I ask because you only have 'recv' in connectivityObserveConnectedSlotAppArmor and replies are automatically allowed.

xavi-garcia-mena replied:
Yes, send is enough.
Done.

jdstrand commented on interfaces/builtin/connectivity_observe_test.go:
```diff
@@ -0,0 +1,83 @@
+	// connected plugs have a non-nil security snippet for apparmor
+	snippet, err = s.iface.ConnectedPlugSnippet(s.plug, s.slot, interfaces.SecurityAppArmor)
+	c.Assert(err, IsNil)
+	c.Assert(snippet, Not(IsNil))
```
Can you verify that something from your ConnectedSlot policy was added? Eg:
```
c.Check(string(snippet), testutil.Contains, "peer=(label=\"snap.connectivity-observe.server\"")
```

xavi-garcia-mena replied:
I've added:
	c.Check(string(snippet), testutil.Contains, "peer=(label=\"snap.server.*\"")
	c.Check(string(snippet), testutil.Contains, "interface=com.ubuntu.connectivity1.NetworkingStatus{,/**}")

Same as above. jdstrand was suggesting: label=\"snap.connectivity-observe.server\", but after running the tests I saw the label was different.

niemeyer commented:
@xavi-garcia-mena This is ready for some love.

alecu commented:
@niemeyer: the proposed connectivity-observe interface gives apps a small r/o api with one enum: "online/offline/connected" and an extra flag that's set when the connection is metered so apps can avoid using a lot of bandwidth with long transfers.

network-observe on the other hand provides r/o access to all of network manager, where much more sensible information is available, like ip addresses, access point names, etc. It's not suitable for all apps, but instead for some system indicators and tools.

niemeyer commented:
@alecu Thanks for the details.

I suggest we split things this way:

* *network-control* - Full network control.
* *network-observe* - Full network observation.
* *network-status* - Access to uncompromising online/offline connectivity status.

These are not tied to a particular service, in that we can add multiple APIs to any of them.

alecu commented:
@xavi-garcia-mena I think gustavo's suggestion looks much better, let's go with network-status as the interface name.

jdstrand commented:
+1 on @niemeyer's suggestion.

xavi-garcia-mena commented:
@alecu @jdstrand  @niemeyer 

I've changed the name of the interface and updated the code as requested.

Cheers,

Xavi

niemeyer commented:
@xavi-garcia-mena Can you please confirm individually that all requests by jdstrand above were addressed, or discuss the ones that need discussion?  Even saying just "Done" on them is enough. It helps both you to make sure nothing was missed, and others to understand you did look through them and addressed or replied.

xavi-garcia-mena commented:
@niemeyer, sure... I've added the replies, sorry for that.
@jdstrand, please take a look more deeply at the tests. You were suggesting a peer label that I saw after running the tests that it was not matching. I don't know if that means I did anything wrong.
Thanks!

xavi-garcia-mena commented:
Thanks @jdstrand for the review and the clarification on the basedeclaration file.
I've fixed the code following your comments.

Regarding the go fmt issue I think it was because I fixed a conflict with the github conflict editor... bad choice.

jdstrand commented:
This needs to be converted to use the new spec mechanism. You might want to talk to @stolowski for details if you have trouble. I suspect this will start making all the tests start to pass. I'll review this PR again once that is done and the tests are passing to verify all the policy is being put in the right places.

xavi-garcia-mena commented:
@jdstrand Hi, yeah, I saw that yesterday... 
I'm already changing the code.
Thanks!

zyga commented:
This is now converted in my brach and @xavi-garcia-mena will pull and push here

xavi-garcia-mena commented:
@jdstrand I've been waiting for the checks and I see that the xenial-ppc64el is failing. I didn't see any failure related to my code, though. 

Now the interface follows the new specs

zyga commented:
Hey @xavi-garcia-mena can you please merge master into this branch?
@niemeyer / @jdstrand can you please re-review this?

xavi-garcia-mena commented:
@zyga, @niemeyer , @jdstrand Master merged into the branch.

zyga commented:
This is failing on (amongst others) ```
----------------------------------------------------------------------
FAIL: overlord_test.go:360: overlordSuite.TestEnsureLoopPrune

overlord_test.go:384:
    c.Assert(st.Change(chg1.ID()), Equals, chg1)
... obtained *state.Change = (*state.Change)(nil)
... expected *state.Change = &state.Change{state:(*state.State)(0x18fa9200), id:"1", kind:"abort", summary:"...", status:0, clean:false, data:state.customData{}, taskIDs:[]string{"1"}, lanes:0, ready:(chan struct {})(0x18f94300), spawnTime:time.Time{sec:63626922395, nsec:637788939, loc:(*time.Location)(0x8ac2bc0)}, readyTime:time.Time{sec:63626922395, nsec:739581000, loc:(*time.Location)(0x8ac2bc0)}}
```

I'm trying to address that in https://github.com/snapcore/snapd/pull/3140

zyga commented on interfaces/builtin/all_test.go:
```diff
@@ -49,10 +49,11 @@ func (s *AllSuite) TestInterfaces(c *C) {
 	c.Check(all, DeepContains, &builtin.LocationControlInterface{})
 	c.Check(all, DeepContains, &builtin.LocationObserveInterface{})
 	c.Check(all, DeepContains, &builtin.LxdSupportInterface{})
```
I re-sorted the file so this is where the extra changes come from.
